### PR TITLE
`src/itx.rs`: Deduplicate w/ generics

### LINKED ISSUE #25: Fix duplicated _tmpl.rs files
The original C code contains a number of `_tmpl.c` files that get compiled once for each active bitdepth. This means that if both 8 and 16 bitdepths are enabled those files get compiled twice. To mimic this in the transpiled Rust, we have transpiled these files twice, e.g. `cdef_tmpl.c` gets transpiled to `cdef_tmpl_8.rs` and `cdef_tmpl_16.rs`.

We should see if we can merge the two Rust files back into a single one. Currently there's a lot of duplicated code between the two versions of a given file, with only relatively small differences between the two in some cases.

### PR DESCRIPTION
**Pull request by thedataking**
First commit emits debug info for assembly on debug builds. This aids testing and debugging. Closes #25.

NOTE to self: need to update the macro docs in `bitdepths.rs` if the overall approach looks good.

### CODE REVIEW

File: src/itx.rs [hunk 1]
```diff
@@ -409,3 +452,803 @@ inv_txfm_fn64!(32, 64, 1);
...
+        let mut x_1 = 0;
+        while x_1 < 4 {
+            let fresh1 = c;
+            c = c.offset(1);
+            *dst.offset(x_1 as isize) = iclip(
+                (*dst.offset(x_1 as isize)).as_::<c_int>() + *fresh1,
+                0 as c_int,
+                bitdepth_max,
+            )
+            .as_::<BD::Pixel>();
```
Comment (kkysen) [inline src/itx.rs:515]: Should be a `bd.iclip_pixel`.

File: src/itx.rs [hunk 2]
```diff
@@ -409,3 +452,803 @@ inv_txfm_fn64!(32, 64, 1);
...
+        assign_itx2_bpc_fn! (c, R,  8, 32, 8 bpc, ssse3);
+        assign_itx2_bpc_fn! (c, R, 16, 32, 8 bpc, ssse3);
+        assign_itx2_bpc_fn! (c, R, 32,  8, 8 bpc, ssse3);
+        assign_itx2_bpc_fn! (c, R, 32, 16, 8 bpc, ssse3);
+        assign_itx2_bpc_fn! (c,    32, 32, 8 bpc, ssse3);
+        assign_itx1_bpc_fn! (c, R, 16, 64, 8 bpc, ssse3);
+        assign_itx1_bpc_fn! (c, R, 32, 64, 8 bpc, ssse3);
+        assign_itx1_bpc_fn! (c, R, 64, 16, 8 bpc, ssse3);
+        assign_itx1_bpc_fn! (c, R, 64, 32, 8 bpc, ssse3);
+        assign_itx1_bpc_fn! (c,    64, 64, 8 bpc, ssse3);
```
Comment (kkysen) [inline src/itx.rs:752]: If you think it's worth it, it'd be nice to have these in the same order as C does.  It makes it easier to check that it's right, and easier to backport changes if new asm functions are added.

Same for the other assigning blocks like this I assume.

**Review by kkysen — CHANGES_REQUESTED** [review]
I haven't thoroughly reviewed the macros and initialization logic (comparing them to C), as you said you looked at that pretty thoroughly already.  I can if you think you could've missed something, though.  Besides that, I had a few other small comments.

File: include/common/bitdepth.rs [hunk 3]
```diff
@@ -375,7 +375,7 @@ impl BitDepth for BitDepth16 {
         self.bitdepth_max
     }
 
-    /// 4 for 10 bits/component.  
+    /// 4 for 10 bits/component.
```
[the remark this answers is not in the document]
Comment (thedataking) [reply include/common/bitdepth.rs:378]: I didn't do this explicitly; I set VSCode to trim trailing whitespace on save. If this gets rendered as Markdown, would it work to make this a list? I think its fairly common for editors to trim whitespace.

```
/// - 4 for 10 bits/component.  
/// - 2 for 12 bits/component.
```

so 

- 4 for 10 bits/component.  
- 2 for 12 bits/component.

File: include/common/bitdepth.rs [hunk 4]
```diff
@@ -561,11 +561,25 @@ macro_rules! bd_fn {
...
         use $crate::include::common::bitdepth::fn_identity;
 
         bd_fn!(fn_identity, $BD, $name, $asm)
     }};
+
+    ($bpc:literal bpc, $name:ident, $asm:ident) => {{
+        use $crate::include::common::bitdepth::fn_identity;
+
+        bd_fn!(fn_identity, $bpc bpc, $name, $asm)
+    }};
```
[the remark this answers is not in the document]
Comment (thedataking) [reply include/common/bitdepth.rs:582]: Yes, I agree, adding a separate `bpc_fn!` macro.

File: src/itx.rs (hunk 2 above)
Comment (thedataking) [reply src/itx.rs:752]: Yes, I agree we should keep it in C order. 

The order you're seeing is from applying the macros step by step to the existing Rust. Somehow that diverged from the order of the C code.

Comment (kkysen) [reply src/itx.rs]: Ahh, that's weird that c2rust did that.  Oh well.

File: include/common/bitdepth.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkysen) [reply include/common/bitdepth.rs]: Hmm, that looks fine for this case, but I don't think it's a good idea for your editor to do that always.  `rustfmt` already trims trailing whitespace in code, but leaves it in doc comments like this where it knows trailing whitespace has an effect, so if you just set vscode to run `rustfmt` on save, that would be better.

[the remark this answers is not in the document]
Comment (randomPoison) [reply include/common/bitdepth.rs]: For the record I don't think it's a good idea to use trailing whitespace for formatting. I also prefer to have my editor trim trailing whitespace (though I think I've been relying on rustfmt for rav1d). We should use normal markdown formatting like a list (as Per suggested).

[the remark this answers is not in the document]
Comment (kkysen) [reply include/common/bitdepth.rs]: Trailing `  ` (double space) is normal markdown for a line break, though.

**Review by randomPoison — APPROVED** [review]
The translation of the C macros looks correct to me. Is there a reason you didn't setup macros for assigning the C fallback functions though? I'm assuming it's correct because what we should have should be the direct result of expanding the macros in C and should have been the same between the 8bpc and 16bpc files, so if you copied it out from one of them then it should be correct here. But it's hard to verify without having the macros setup the same way.

**Comment by kkysen** [comment]
We can later change the fallbacks to return `Self`, which will check that we initializedeverything.  So I'm less worried about them vs the asm ones.

**Comment by thedataking** [comment]
> Is there a reason you didn't setup macros for assigning the C fallback functions though?

Derp, I forgot to push that change; sorry about that .

**Review by randomPoison — APPROVED** [review]
Updated macros for initializing the fallback fns look good too 👍 Thanks for adding those.

**Pull request merged**